From cc9338e194df4b2f36d5f98e80347e41cf874fba Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:45 -0400
Subject: [ethernet] qlcnic: Fix bug in TX statistics

Message-id: <1408626422-3941-62-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88607
O-Subject: [PATCH RHEL 7.1 061/138] qlcnic: Fix bug in TX statistics
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish Chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 1ac6762a0b8f5f2dc35ea869a08da25c68d7a8ba

o Driver was not updating TX stats so it was not populating
  statistics in `ifconfig` command output.

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h b/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
index 35d4876..4f7bdfd 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic.h
@@ -1720,6 +1720,7 @@ int qlcnic_83xx_init_mailbox_work(struct qlcnic_adapter *);
 void qlcnic_83xx_detach_mailbox_work(struct qlcnic_adapter *);
 void qlcnic_83xx_reinit_mbx_work(struct qlcnic_mailbox *mbx);
 void qlcnic_83xx_free_mailbox(struct qlcnic_mailbox *mbx);
+void qlcnic_update_stats(struct qlcnic_adapter *);
 
 /* Adapter hardware abstraction */
 struct qlcnic_hardware_ops {
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
index 45fa6ef..a13b278 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ethtool.c
@@ -1270,7 +1270,7 @@ static u64 *qlcnic_fill_stats(u64 *data, void *stats, int type)
  return data;
 }
 
-static void qlcnic_update_stats(struct qlcnic_adapter *adapter)
+void qlcnic_update_stats(struct qlcnic_adapter *adapter)
 {
  struct qlcnic_host_tx_ring *tx_ring;
  int ring;
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
index 49bf425..4fdfdb2 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_main.c
@@ -2774,6 +2774,9 @@ static struct net_device_stats *qlcnic_get_stats(struct net_device *netdev)
  struct qlcnic_adapter *adapter = netdev_priv(netdev);
  struct net_device_stats *stats = &netdev->stats;
 
+ if (test_bit(__QLCNIC_DEV_UP, &adapter->state))
+  qlcnic_update_stats(adapter);
+
  stats->rx_packets = adapter->stats.rx_pkts + adapter->stats.lro_pkts;
  stats->tx_packets = adapter->stats.xmitfinished;
  stats->rx_bytes = adapter->stats.rxbytes + adapter->stats.lrobytes;
-- 
1.7.1